### PR TITLE
CI: format isort output for detection by azure pipelines

### LINKED ISSUE #27179: CI: Format isort output to be detected as Azure pipelines errors
`isort` is a tool to validate the order of the imports in Python code. We are using it in the CI to make sure all pandas imports follow a defined standard. It is implemented in `ci/code_checks.sh`.

When `isort` detects imports incorrectly sorted, it finishes with an exit code different than 0, and that makes the CI red. Besides that, it'd be useful that in Azure Pipelines (the CI system where `isort` runs) the output messages of `isort` were detected as errors.

Some tools provide an option to specify the output format, like `flake8`, but `isort` doesn't (see https://github.com/timothycrosley/isort/issues/961).

What we can do in this case is to modify the output with bash pipelines. We are already doing that in the `invgrep` function in `ci/code_checks.sh`.

What needs to be done is:
- Change the format of `isort` output to be: `##vso[task.logissue type=error;sourcepath=%(path)s]%(text)s"
- Make sure that the resulting command still has an exit code of `0` if no issues are found, and the exit code of `isort` is there are. e.g. the exit code of `isort -c foo.py | awk -F" " '{ print $2 }'` is 0, the exit code of `awk` even if errors are found (the exit code of `isort` is lost in the pipeline).

Note that the error message of `isort` is not delimited in a way that can be easily parsed. But since the error message does not change, it should be possible to extract the name of the file removing the number of characters of the text around it (see https://github.com/timothycrosley/isort/blob/develop/isort/isort.py#L122).

### PR DESCRIPTION
**Pull request by leeyspaul**
- [ ] closes #27179 
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: ci/code_checks.sh [hunk 1]
```diff
@@ -109,7 +109,7 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
 
     # Imports - Check formatting using isort see setup.cfg for settings
     MSG='Check import format using isort ' ; echo $MSG
-    isort --recursive --check-only pandas asv_bench
+    isort --recursive --check-only pandas asv_bench | awk -F ":" '##vso[task.logissue type=error;sourcepath=%(path)s]%(text)s'
```
Comment (datapythonista) [inline ci/code_checks.sh]: Looks good, but we just want to change the format when the environment variable `AZURE` is set. You can see how this is done for flake8. Sorry it wasn't clear in the issue.

Can you also change the order of an import to see this failing in the CI please? So we can see it working before we merge. Thanks!

Comment (leeyspaul) [reply ci/code_checks.sh]: yes, thank you for being so helpful!

Comment (WillAyd) [reply ci/code_checks.sh]: Does this still grab the return code from `isort`?

**Comment by leeyspaul** [comment]
@datapythonista sorry I'm a bit new to bash scripting. It's definitely catching the import error but it is ignoring the exception as shown here: 

```Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>
BrokenPipeError: [Errno 32] Broken pipe
Check import format using isort DONE
##[error]Bash exited with code '120'.```

**Comment by leeyspaul** [comment]
pinging for review! Spent some time on it to understand `awk`. @datapythonista  thank you!

File: ci/code_checks.sh [hunk 2]
```diff
@@ -109,9 +109,12 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
 
     # Imports - Check formatting using isort see setup.cfg for settings
     MSG='Check import format using isort ' ; echo $MSG
-    isort --recursive --check-only pandas asv_bench
+    if [[ "$AZURE" == "true" ]]; then
+        isort --recursive --check-only pandas asv_bench | awk -F ':' '{print $0"\n" "##vso[task.logissue type=error;sourcepath=%(path)s]%(text)s"}'
+    else
+        isort --recursive --check-only pandas asv_bench | awk -F ':' '{print}'
```
Comment (datapythonista) [inline ci/code_checks.sh]: I don't think we need `awk` in this case, do we?

File: ci/code_checks.sh [hunk 3]
```diff
@@ -109,9 +109,12 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
 
     # Imports - Check formatting using isort see setup.cfg for settings
     MSG='Check import format using isort ' ; echo $MSG
-    isort --recursive --check-only pandas asv_bench
+    if [[ "$AZURE" == "true" ]]; then
+        isort --recursive --check-only pandas asv_bench | awk -F ':' '{print $0"\n" "##vso[task.logissue type=error;sourcepath=%(path)s]%(text)s"}'
```
Comment (datapythonista) [inline ci/code_checks.sh]: I think the return code `$?` we are checking later to see if `isort` was successful or not, will actually be the one of awk with this, which will always be 0.

This means that if the imports are not correctly sorted, we won't detect them in the CI.

Can you start by swapping the order of two imports in pandas in this PR, so we know if this is working or not? We'll remove them once the PR is approved.

File: ci/code_checks.sh [hunk 4]
```diff
@@ -109,9 +109,12 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
...
+    else
+        isort --recursive --check-only pandas asv_bench | awk -F ':' '{print}'
+    fi
     RET=$(($RET + $?)) ; echo $MSG "DONE"
-
```
Comment (datapythonista) [inline ci/code_checks.sh]: in general try to avoid unrelated changes, even if they are small like this.

**Review by datapythonista — CHANGES_REQUESTED** [review]
Thanks @leeyspaul, getting closer, still some comments.

File: ci/code_checks.sh (hunk 3 above)
Comment (leeyspaul) [reply ci/code_checks.sh]: Yes, I will push up the badly sorted imports

File: ci/code_checks.sh (hunk 4 above)
Comment (leeyspaul) [reply ci/code_checks.sh]: Undid the change!

File: ci/code_checks.sh (hunk 3 above)
Comment (leeyspaul) [reply ci/code_checks.sh]: here is the output! ```Check import format using isort
ERROR: /home/vsts/work/1/s/pandas/plotting/_matplotlib/core.py Imports are incorrectly sorted.
##[error]%(path)s(,): error : %(text)s
Check import format using isort DONE
##[error]Bash exited with code '1'.
##[section]Finishing: Linting```

**Comment by datapythonista** [comment]
Seems like `awk` is keeping the exit status of isort, I guess it's because of the `set -o pipefail` in the function at the top of the function, we'll have to fix that.

But the first things would probably be fixing the output, from:
```
ERROR: /home/vsts/work/1/s/pandas/plotting/_matplotlib/core.py 
##[error]%(path)s(,): error : %(text)s
```
to
```
##vso[task.logissue type=error;sourcepath=/home/vsts/work/1/s/pandas/plotting/_matplotlib/core.py] Imports are incorrectly sorted.
```

Thanks!

**Comment by leeyspaul** [comment]
ok! I think I'm getting closer now. @datapythonista could you check out the failure? I've pasted below what it's returning for ease.

```
Check import format using isort
##[error]pandas/plotting/_matplotlib/core.py(,): error :   Imports are incorrectly sorted.Check import format using isort DONE
##[error]Bash exited with code '1'.
##[section]Finishing: Linting
```

I added the ORS as the error message was moving to a new line with print. If there is a better way to do this do let me know, thank you!

**Comment by datapythonista** [comment]
is there a reason to not use the `##vso` code I suggested? I think is what we use in the other cases, if there is no reason better to keep consistent.

**Comment by leeyspaul** [comment]
> is there a reason to not use the `##vso` code I suggested? I think is what we use in the other cases, if there is no reason better to keep consistent.

er, I didn't mean to remove anything. I'm not too familiar on vso codes but I read up on logging here: https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/authoring/commands.md

Could you clarify what you mean by the vso code and I will change it back, or point me to the docs. Thank you!

**Comment by datapythonista** [comment]
I refer to what is suggested in https://github.com/pandas-dev/pandas/pull/27334#issuecomment-515995494 instead of using `##[error]`

**Comment by leeyspaul** [comment]
ah right. So oddly it seems to be getting converted from `##vso` to `##[error]` though the awk line I've used is `isort --recursive --check-only pandas asv_bench | awk '{ORS=""} {print "##vso[task.logissue type=error;sourcepath="$2";]"} {$1=$2=""; print $0}'`.

I might be missing something obvious here

**Comment by datapythonista** [comment]
ah, sorry, didn't see in the code you were already using the `##vso`. I guess that's some azure magic then, but will be applied to all the other commands too, so it's ok.

I'd prefer to remove the `awk` and leave the original call when not using azure. Will be misleading if the errors are different when people call `isort` directly than with our script I think.

And the thing that concerns me is the `set -o pipefail` inside the if at the beginning. If eventually we change that code, your code will fail. And that doesn't seem right, since both things are unrelated, and shouldn't be using side-effects of each other. I'd start by adding `set +o pipefail` after the grep in line 39. That should already be there, will avoid the side effects, and will make your code fail. After that, you can wrap your code with the same `set -o pipefail` and `set +o pipefail`, and that should fix it again.

Does this make sense?

**Comment by leeyspaul** [comment]
> ah, sorry, didn't see in the code you were already using the `##vso`. I guess that's some azure magic then, but will be applied to all the other commands too, so it's ok.
> 
> I'd prefer to remove the `awk` and leave the original call when not using azure. Will be misleading if the errors are different when people call `isort` directly than with our script I think.
> 
> And the thing that concerns me is the `set -o pipefail` inside the if at the beginning. If eventually we change that code, your code will fail. And that doesn't seem right, since both things are unrelated, and shouldn't be using side-effects of each other. I'd start by adding `set +o pipefail` after the grep in line 39. That should already be there, will avoid the side effects, and will make your code fail. After that, you can wrap your code with the same `set -o pipefail` and `set +o pipefail`, and that should fix it again.
> 
> Does this make sense?

I think so, I'll take a look at this tonight! Thanks for the info. : )

**Comment by WillAyd** [comment]
@leeyspaul can you merge master again? Old builds are no longer available

**Comment by WillAyd** [comment]
Great! No worries on timing - not in any rush. Just wanted to make sure this was something that you were still looking at, so thanks for confirming

**Comment by jreback** [comment]
@leeyspaul can you merge master and update

**Comment by jbrockmendel** [comment]
@leeyspaul can you rebase

**Comment by alimcmaster1** [comment]
If we dont here from @leeyspaul I can take this on

**Comment by leeyspaul** [comment]
@alimcmaster1 feel free to take over. Thanks!

**Comment by alimcmaster1** [comment]
Restarted this over here: https://github.com/pandas-dev/pandas/pull/29654

Closing this for now

**Comment by datapythonista** [comment]
I'm moving the Checks build to actions, and I think it only supports the `##[error]` formatting, not the one with file, line number...

**Comment by alimcmaster1** [comment]
Ahh so there probably isn’t much value in this at all? I can close my PR if so

**Comment by datapythonista** [comment]
There is surely value on letting the CI know what's an error. Just the format will probably be different.
